cdap-proto/src/test/java/co/cask/cdap/proto/id/EntityIdTest.java

```diff
@@ -268,6 +269,51 @@ public void testCompatibility() {
     }
   }
 
+
+  @Test
+  public void testFromMetadataEntity() {
+    ApplicationId applicationId = new ApplicationId("testNs", "app1");
+    Assert.assertEquals(applicationId, EntityId.fromMetadataEntity(applicationId.toMetadataEntity()));
+
+    MetadataEntity metadataEntity = MetadataEntity.ofNamespace("testNs").appendAsType(MetadataEntity.APPLICATION,
+                                                                                      "app1");
+    Assert.assertEquals(applicationId, EntityId.fromMetadataEntity(metadataEntity));
+
+    metadataEntity = MetadataEntity.ofDataset("testNs", "testDs").appendAsType("field", "testField");
+    try {
+      EntityId.fromMetadataEntity(metadataEntity);
+      Assert.fail("Should have failed to get create an EntityId from MetadataEntity");
+    } catch (IllegalArgumentException e) {
+      // expected
+    }
+
+    ProgramId programId = new ProgramId(applicationId, ProgramType.WORKER, "testWorker");
+    Assert.assertEquals(programId, EntityId.fromMetadataEntity(programId.toMetadataEntity()));
+
+    ArtifactId artifactId = new ArtifactId("testNs", "testArtifact-1.0.0.jar");
+    Assert.assertEquals(artifactId, EntityId.fromMetadataEntity(artifactId.toMetadataEntity()));
+  }
```

Comment:
test MetadataEntity.ofNamespace(a).append("non", "sense").appendAsType("dataset", "myds").

what is the expected outcome?

Reply:
Fixed with checking.

Reply:
Also updated to use Builder and do check during the build (last stage)